kuma/api/tasks.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import unicode_literals
+
+import json
+
+import boto3
+from celery import task
+from django.conf import settings
+
+from kuma.core.utils import chunked
+from kuma.wiki.models import Document
+
+from .v1.views import document_api_data, get_content_based_redirect, get_s3_key
+
+
+S3_MAX_KEYS_PER_DELETE = 1000
+
+
+def get_s3_bucket():
```

Comment:
Every single call to `publish()` or `unpublish()` will trigger this function which will create a new HTTPS connection object I think. 
I haven't actually tested it yet but it should be possible to see if you set the boto3 logger to `DEBUG`.

Perhaps that's fine and might only matter those rare few times you manually run `./manage.py publish --all` which you're doing interactively anyway so you can just restart it if you get weird network connection errors.

Reply:
Sorry, I wanted to comment, but just realized that I hadn't yet! Yes, this is a great point, and worth looking into more! I'm not sure what boto3 does. I think the use of `boto3` in the line of code `boto3.resource('s3')` forces the use of the default session object. From some quick poking around on the web, that default session object may provide a connection pool of 10 connections (https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html), but I'm surprised that I can't find an obvious piece of boto3 documentation to confirm or deny that. I'm going to see if I can gather something from the `boto3` source code, but I've looked into the source code before and it's not simple to navigate and understand.

I did see that boto3 recommends a separate session for each thread (https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html#multithreading-multiprocessing), but since we're not running any threads (just processes) in our `gunicorn` set-up, I think we're OK there.

Reply:
After some digging into the source code and running an experiment in the shell, I think I should be calling `boto3.resource('s3')` only once. Every time it's called, although it's reusing the default session, it's creating a new resource object with an underlying new client with its own connection pool of up to 10 connections? Frustrating how bad the documentation is on this point. I think I need to modify this code.

Reply:
Quick story; the reason this stuff makes me nervous was because when I built symbols.mozilla.org it relies on doing a LOT of HEAD requests to S3. And it's perfectly normal, and common, that the S3 key you're asking for does NOT exist. Like 9 out of 10 times, we expect it to not exist. 
So doing `mybucket.head_object({'Key': key})` would ultimately trigger a python exception in boto3. So, I did something like this:

```python
try:
    mybucket.head_object({'Key': key})
    return "Yes, it exists!"
except botocore.exceptions.ClientError as exception:
    if exception.response["Error"]["Code"] == "NoSuchKey":
        return "Bummer, it does not exist"
    raise  # Any other unexpected exceptions
```

Makes sense, right? 
But when starting all of this up in Staging for the first time I noticed the Django server performance sucked. It could only handle a couple of requests per second!
I turned over every stone and started to suspect everything and anything. 

It was not until I upped the logging level that boto3 uses to DEBUG and that I actually started studying all the verbose logging. 

What was happening was that ANY Python exception would cause the connection pool to drain. Even if I, in a higher level, handled the exception. 
So every time it happened, all the connections were lost and the whole HTTP connection (including initial SSL negotiation) had to start over. It was fast (<1 seconds) but the service had to handle ~100 requests/second and it bottlednecked on boto3 creating brand new HTTP connections way too often.

By the way, the solution was crazy but worked. Instead of `mybucket.head_object()` I use:
```
for found in mybucket.list_objects_v2({'Prefix': key}):
    return "Yes it exists!"
return "Bummer it does not exist"
```
I.e. doing a prefix search with the prefix being the whole key name.